### ISSUE TITLE
Merger role

### ISSUE THREAD

**Issue by jacg** [next-exp/IC#509]
Merging approved PRs isn't difficult, but it does require a minute or two to gather the PR number, the author and reviewer comments and identities, in order to prepare the merge commit properly.

Sometimes this cannot be done immediately after approval, as Travis checks must be awaited. All in all, this sometimes (too often) results in approved PRs going unmerged for weeks, which not only unnecessarily delays the inclusion of useful work, but also allows merge conflicts to creep in.

In summary, we want to avoid this.

I propose that we give a junior developer (@carmenromo ?, Ryan?) the role of merger. This would involve looking at the PR list every day, and merging any approved PRs (except those marked as *do not merge*).

This would perform a useful service and allow the person in the role to keep abreast of what is going on in the project.

**Comment by carmenromo** [comment]
Perfect! Also Ryan is starting on IC these days so we can do this work together

**Comment by jjgomezcadenas** [comment]
Terrific idea. I hereby proclaim Carmen and Ryan as official IC mergers (Carmen, try as soon as possible with the next merge UNDER GUIDANCE OF A SENIOR, best if @jmbenlloch or the master @jacg himself).

**Comment by jacg** [comment]
The procedure is:

In the following I assume that you have a remote called `nextic` pointing at the central IC repository.

I also assume that you already have a remote configured for the author in question. If not, you will have to create one first.

1. Fetch `nextic/master`. [CLI: `git fetch nextic master` Magit: `f o nextic RET master RET`]
1. Fetch the branch of the approved PR. [CLI: `git fetch <author's remote> <PR branchname>` Magit `f o <author's remote> RET <PR branchname> RET`]
1. Create and check out a local branch on top of the PR branch. [CLI: `git checkout -b <PR branchname> nextic/<PR branchname>` Magit: `b c nexic RET <PR branchname> RET RET`]
1. Rebase the branch of the approved PR onto `nextic/master` [CLI: `git rebase nextic/master` Magit `r e nextic/master RET`]
    + If there are conflicts, ask the author to resolve them, unless they are obvious.
1. Move your `master` branch to `nextic/master` [CLI: `git branch -f nextic/master` Magit `b x master RET nextic/master`]
1. Checkout your local `master` [CLI: `git checkout master` Magit: `b b master RET`]
1. Merge the PR branch into your local `master`, making sure that the merge commit conforms to our requirements:
    + Disallow *fast forward* merging: we want an explicit merge commit for each PR.
    + Make sure that the merge commit author is the person who formaly approved the PR
    + The title of the merge commit must start with the PR number, and be followed by the title of the PR
    + The body of the commit must contain 2 sections
       1. The first must have the heading `[author: <author's id>]` and contain the description of the PR. This is usually the whole first comment in the PR
       2. The second must have the heading `[reviewer: <approver's id>]` and contain the approver's comment, usually just a verbatim copy of the comment the reviever left on GitHub.

       Here are the steps needed to make this happen:

    1. Create a *no fast forward* merge commit. [CLI: `git merge --edit --no-ff <PR branchname>` Magit: `m -n e <PR branchname> RET`]

    1. Edit the commit message as described above so that it has the following format:

        <PR number> <PR title>

        [author: <author id>]

        PR description

        [reviewer: <reviewer id>]

        Reviewer comment

    1. Ensure that Git records the reviewer of the PR as the author of the merge commit. [CLI: `git commit --amend --author=<reviewer's id>` Magit: `c =A <reviewer's id> RET w C-c C-c`]

1. If the top comment of the PR contains any WIP markers or caveats or warnings about the work being incomplete, these should (usually) no longer be valid once the PR is accepted and merged, so please remove them by editing the top comment (after having made sure that they really are no longer relevant).

1. Push the merge commit to the central repository. [CLI: `git push nextic master` Magit: `P e nextic/master`]

(Usually, the Magit commands will involve less typing than shown here, because Magit and Helm will be able to guess what you are doing most of the time and pick defaults or proved shortcuts, but I'm being conservatve and thus long-winded in my desciption.)

If this idea is approved, then you will need to be given write permission to the central repo, before you will be able to do the final step. If not, well, here is a recipe for merging PR according to our project standards.

I would love to guide you through your first merge, but given that I am a few thousand km away, and this is something that is best done with peras ... you should probaby grab someone local to walk you through it.

**Comment by jacg** [comment]
Oh, I forgot to mention, you can practise all you want before doing it for real, by simply not doing the last step (pushing to the central repo). Just pick any open PR and pretend that it has been approved. You will have to invent a reviewer and an approval message. Go through all the steps except the very last one. Then simply reset your `master` branch to `nextic/master` and (maybe) delete your local version of the PR branch, and you're back to square one. If you mess up, it really doesn't matter (as long as you don't push to `nextic/master`) as you can always forget it all by resetting `master` to `nextic/master` and deleting your local branch.

Go for it. Have fun!

**Comment by jacg** [comment]
One more detail, when changing the commit author, sometimes the user's GitHub handle works, sometimes I need to use the full name as it appears in the email address known to GitHub. I haven't got to the bottom of this little mystery.

**Comment by carmenromo** [comment]
Wow thank you so much! I will ask @jmbenlloch to help me if he can, although he taught me some weeks ago how to do a PR and merge it.

**Comment by jacg** [comment]
One more thing about practising: instead of pushing to `nextic/master` you could push to your own IC fork on GitHub; that way someone (me for example) could give you feedback remotely. But it would, of course, be simplest if you just go through the steps with someone looking over your shoulder.

**Comment by carmenromo** [comment]
ok, understood

**Comment by jacg** [comment]
Oh, and stay well clear of #435! It's been approved, but so many moons ago, that getting this one rebased and all the new development that has happened since then, adapted to the style therein, will be a Herculean task of technical debt repayment which will form a major part of the upcoming sprint. Don't go there. Here be dragons!

**Comment by jacg** [comment]
Given that @jjgomezcadenas has officially approved this idea and appointed mergers, I hereby officially close this issue.

I'm in the process of creating a Kanban board, to help us mitigate the workflow disaster which has evolved over the last months. On there I have created a *merge* column, from which the mergers should be able to pick up their tasks for this role. Hopefully there will be a way to automatically notify the mergers when some item gets moved to that column.

**Comment by jacg** [comment]
Apologies, I made a mistake in the instructions I gave above: In Magit a `--no-fast-forward` merge is done with `-n` rather than `-f`.

I have edited the comment in question to fix the error.
